src/block.rs

```diff
@@ -1,88 +1,198 @@
 //! Block validation
-use crate::cid::{Cid, Codec as RawCodec};
-use crate::error::{BlockError, IpldError};
-use crate::hash::{digest, Hash};
+use crate::cid::Cid;
```

Comment:
Here is where my proposal of the Block API (https://github.com/ipfs-rust/rust-ipld/issues/14) comes in. I've created a small demo application which used the proposed API (some details changed, but the idea is the same). The full implementation of that API can be found here: https://github.com/vmx/rust-dagbuilder/blob/f7f0c00b96074f6aa6a5aeef0b327b35d73dea72/src/block.rs

More information is available on the [Block API issue](https://github.com/ipfs-rust/rust-ipld/issues/14#issuecomment-624324812).

I hope that we can use my proposed API (of course after further discussions) instead.

My code currently uses a modified libipld, tomorrow I'll try to base it on this PR, then I can also say more about the encoder/decoder changes.

Reply:
One thing I don't like is this:
```rust
Box::<dyn Codec>::from(self.codec).decode(&raw)
```
There needs to be a trait Code:
```rust
trait Code {
    fn encode(&self) {
        match self {
            Self::Cbor => Cbor::encode(),
        }
    }
    fn decode(&self)
}

impl Code for Codec {}
```
same in multihash there needs to be a trait Code:
```rust
trait Code {
    fn digest()
}
```

Reply:
encode and decode should return a reference, the caller can clone if ownership is required.

the block api also doesn't support encoding structs directly, Ipld is required as an intermediate representation.

I don't see necessarily that the two api's are conflicting, they have different use-cases.

Reply:
> the block api also doesn't support encoding structs directly, Ipld is required as an intermediate representation.

That's something I haven't thought about yet. But it's a good point. So far I've seen the `Ipld` enum as the central piece everything should center around. But perhaps we can get away with using native types. I need to thing about that.

Also perhaps the block api can be extended to also support native types.